job/job.go

```diff
@@ -0,0 +1,188 @@
+package job
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strings"
+	"time"
+)
+
+type Job struct {
+	Id string `json:"id"`
+}
+
+type Status struct {
+	Status   string `json:"status"`
+	Location string `json:"location"` // Location is the unauthenticated temporary (60 secs) url used to download data
+}
+
+func CreateJob(url, authToken string, jobConfig string) (*Job, error) {
+
+	JobResponse := Job{}
+
+	payload := strings.NewReader(jobConfig)
+
+	req, _ := http.NewRequest("POST", url, payload)
+
+	req.Header.Add("authorization", fmt.Sprintf("Bearer %s", authToken))
+	req.Header.Add("content-type", "application/json")
+
+	res, err := http.DefaultClient.Do(req)
+	if res.StatusCode != http.StatusOK {
+		err = fmt.Errorf("expected %d, got %d: %s", http.StatusOK, res.StatusCode, res.Body)
+		return nil, err
+	}
+
+	defer res.Body.Close()
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	JobResponse, err = handleData(body)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("Created job %s", JobResponse.Id)
+
+	return &JobResponse, nil
+}
+
+func handleData(body []byte) (Job, error) {
+	var job = Job{}
+	err := json.Unmarshal(body, &job)
+
+	return job, err
+}
+
+func WaitForJobCompletion(url string, authToken string, job Job) (resultURL *string, err error) {
+
+	jobStatusUrl := fmt.Sprintf("%s/%s", url, job.Id)
+
+	var status Status
+	var body []byte
+
+	for {
+		body, err = requestAndHandleData("GET", jobStatusUrl, authToken, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		err := json.Unmarshal(body, &status)
+		if err != nil {
+			return nil, err
+		}
+
+		log.Printf("Job: %s, Status: %s", job.Id, status.Status)
+		if status.Status == "completed" {
+
+			return &status.Location, nil
+		}
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func GetUserExport(resultURL string) (io.ReadCloser, error) {
+
+	client := http.Client{Timeout: 1 * time.Minute}
+
+	req, err := http.NewRequest("GET", resultURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Accept-Encoding", "gzip")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		err = fmt.Errorf("expected %d, got %d", http.StatusOK, resp.StatusCode)
+		return nil, err
+	}
+
+	return resp.Body, nil
+}
+
+func UploadUserExportToS3(awsClient s3iface.S3API, readCloser io.ReadCloser, bucket, key string) (*s3.PutObjectOutput, error) {
+
+	reader, err := gzip.NewReader(readCloser)
+	if err != nil {
+		return nil, fmt.Errorf("error during Gunzip operation while uploading %s:\n%s", key, err)
+	}
+
+	body, err := ioutil.ReadAll(reader)
```

Comment:
I may be wrong but probably you don't need to read this all and then re-wrap it in another reader, unless I'm missing something you could use the (gzip) reader directly in the `Body` field of the `s3.PutObjectInput` struct (still wrapped using the `aws.ReadSeekCloser()` function).

Reply:
Unfortunately no.  I think it may be something to do with content-length not being present which results in a `501`.  In any case, the reader returned from the gzip package does not work in `aws.ReadSeekCloser`

Reply:
What? Really? I though an `io.Reader` is an `io.Reader` so it should just work in `aws. ReadSeekCloser`, weird.

Hopefully the return value from `gzip.NewReader()` would work with `aws.ReadSeekCloser()` ☺️